downloaders: add ACL Anthology downloader

mar4th3 opened the pull request:
Hi! 

This is related to #417 

I finally managed to find some time to take a crack at writing a custom downloader.

Maybe you'd like to have a look?

The main questions that remain is: how do I test if this is working with `papis add`?

Thanks!

alexfikl commented:
> The main questions that remain is: how do I test if this is working with `papis add`?

I'm not quite sure how you set up your dev environment, but I usually have a virtual environment set up and then install `papis` in editable mode
```
python -m pip install -e '.[develop]'
```
which also installs the entry points for the plugins (like the downloaders). Then you can just do
```
papis add https://acl.org/some-id
```
and see if it correctly pics up the downloader you added and all the metadata.

alexfikl commented on HACKING.md:
```diff
@@ -102,27 +98,100 @@ see the arXiv importers in `arxiv.py`. To add a new importer
-similarly to an importer.
+from a remote location. They can be implemented in a similar way.
+
+As a example we implement a custom downloader for the [ACL Anthology](https://aclanthology.org/).
```
The example you added here is great! However it reads more like a tutorial, so maybe it should be in the docs instead?

The docs for the plugin architecture are pretty nonexistent at the moment
https://github.com/papis/papis/blob/2fe163de082b883f804b19ed76d756b02f05dfdb/doc/source/plugins.rst
but we can start something in there.

alexfikl commented on setup.py:
```diff
@@ -4,11 +4,11 @@
-import glob
 
-from setuptools import setup, find_packages
+from setuptools import find_packages, setup
```
I'm guessing most of these changes are from a formatter (black?). We don't really enforce one at the moment, so can you revert these changes?

alexfikl commented on papis/downloaders/acl.py:
```diff
@@ -0,0 +1,89 @@
+
+
+if __name__ == "__main__":
+    TEST_CASES = [
+        (
+            "https://aclanthology.org/N04-1001",
+            {"acl_anthology_id": "N04-1001"},
+        ),
+        (
+            "https://aclanthology.org/2021.naacl-main.208/",
+            {"acl_anthology_id": "2021.naacl-main.208"},
+        ),
+    ]
+    for url, y_true in TEST_CASES:
+        d = Downloader(url)
+        d.fetch_data()
+
+        print(f"Test: {url}")
+        print("Fetched:")
+        print(f"\t {d.ctx.data}")
+        assert d.ctx.data["acl_anthology_id"] == y_true["acl_anthology_id"]
+        print("\n\n")
```
These could go into `tests/downloaders`. There's some existing examples in there, but let me know if you need any help setting them up!

alexfikl reviewed (commented):
The downloader itself looks good to me! (Haven't tested it yet though)

I've added some high level comments for the tests and formatting, but those are just moving things around, so hopefully not too hard to clean up.

mar4th3 commented:
Hi!

I finally managed to find some time for this. I have updated the code following your comments. Could you please take a look again?

I have added a `test_acl.py`. The tests are failing because the resource files are not created:

```
E       FileNotFoundError: [Errno 2] No such file or directory: '/home/ele/Projects/papis/tests/downloaders/resources/ACL_2021.naacl-main.208.html'
```

Stemming from `get_remote_resource`. Any idea what I'm doing wrong?

Thanks!

alexfikl commented:
I'll have to look a bit over the docs, but overall this looks good to me! :heart: 

Do the parsed results in the JSON resource files look good to you?

alexfikl commented:
> > Yeah, that part is a bit undocumented, so I added the missing resources in a few commits.
> 
> Thanks for the cllarification! What do you think about adding a final section into `plugins.rst` about testing?

That's definitely a good idea, but I think we should put it in a separate PR. I have some ideas for simplifying the workflow there and I'll take a crack at it over the weekend!

> Yeah LGTM!

Awesome! Whenever you're ready to take this out of draft status, we can merge it in!

mar4th3 commented:
> By the way, you might need to merge main (or rebase, if that's your preference), since this branch seems to be quite a bit behind. I imagine Github will complain once this is not a draft anymore 😁

I've changed the PR status and no complaints from Github :) I noticed that there are a couple more changes in the PR on other files (e.g. `tui/widgets/diff.py`) but I guess this  is something you introduced.

alexfikl commented:
> I've changed the PR status and no complaints from Github :)

Yeah, everything seems ok, so in it goes! Thank you very much for all your work on this! :rocket:

> I noticed that there are a couple more changes in the PR on other files (e.g. `tui/widgets/diff.py`) but I guess this is something you introduced.

That was me! I had some local errors from `flake8`, not sure what that was about, but it looks ok on the CI!
